File: problems/Java/103. Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) return res;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean ldr = true;
        
        while (!q.isEmpty()) {
            int sz = q.size();
            LinkedList<Integer> lvl = new LinkedList<>();
            
            for (int i = 0; i < sz; i++) {
                TreeNode cur = q.poll();
                if (ldr) {
                    lvl.addLast(cur.val);
                } else {
                    lvl.addFirst(cur.val);
                }
                
                if (cur.left != null) q.add(cur.left);
                if (cur.right != null) q.add(cur.right);
            }
            res.add(lvl);
            ldr = !ldr;
        }
        return res;
    }
}

File: problems/C++/103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root) return res;
        
        queue<TreeNode*> q;
        q.push(root);
        bool ldr = true;
        
        while (!q.empty()) {
            int sz = q.size();
            vector<int> row(sz);
            
            for (int i = 0; i < sz; ++i) {
                TreeNode* cur = q.front();
                q.pop();
                
                int idx = ldr ? i : (sz - 1 - i);
                row[idx] = cur->val;
                
                if (cur->left) q.push(cur->left);
                if (cur->right) q.push(cur->right);
            }
            res.push_back(row);
            ldr = !ldr;
        }
        return res;
    }
};